### PR TITLE
Add options for overriding SPIRV denorm settings.

### PR DESCRIPTION
**Pull request by perlfu**
Add fp32DenormalMode option to pipeline shader options.
Add command line override -fp32-denormal-mode, which can be auto, ftz or preserve.

### CODE REVIEW

**Comment by trenouf** [comment]
What is the intended use case for these? If it is for your experimentation, fine. But I wouldn't want more command line options to be added that are used with app detect in the driver. For that, we should have new option bits in the shader or pipeline info passed in to LLPC.

**Comment by perlfu** [comment]
> What is the intended use case for these? If it is for your experimentation, fine. But I wouldn't want more command line options to be added that are used with app detect in the driver. For that, we should have new option bits in the shader or pipeline info passed in to LLPC.

Might be needed as an app detect, so I will rework.  Do we have anything in the options bits at the moment that effects SPIRV parsing?

**Comment by amdrexu** [comment]
> Do we have anything in the options bits at the moment that effects SPIRV parsing?

We do have options PipelineShaderOptions and PipelineOptions in vkgcDefs.h. But I think they are passed to LGC or are used in llpcCompiler.cpp. Such options are not passed to SPIR-V reader or are not used in lower passes.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by perlfu** [comment]
> > Do we have anything in the options bits at the moment that effects SPIRV parsing?
> 
> We do have options PipelineShaderOptions and PipelineOptions in vkgcDefs.h. But I think they are passed to LGC or are used in llpcCompiler.cpp. Such options are not passed to SPIR-V reader or are not used in lower passes.

Would it be acceptable to change readSpirv to take PipelineShaderOptions then?  Or does this break something in the intended abstraction?

**Comment by amdrexu** [comment]
> Would it be acceptable to change readSpirv to take PipelineShaderOptions then? Or does this break something in the intended abstraction?

I think it is acceptable. Theoretically, those options should either control SPIR-V to LLVM translation or are used by LGC in shader generation.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit b195b54109772ad8fd75a553eb4470f571961c54
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 3cabf261a5df65b35bb702853572c46d23d60acf
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 57b69a2a8caa0bb25a4536c233f5753a0ff544ef
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4010c07e154c19fcee1773b059e3c054e0a6e058
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

File: include/vkgcDefs.h [hunk 1]
```diff
@@ -74,6 +74,7 @@
 //* %Version History
 //* | %Version | Change Description                                                                                    |
 //* | -------- | ----------------------------------------------------------------------------------------------------- |
+//* |     40.4 | Added fp32DenormalMode in PipelineShaderOptions to allow overriding SPIRV denormal settings           |
```
Comment (amdrexu) [inline include/vkgcDefs.h:77]: SPIRV -> SPIR-V

File: llpc/context/llpcPipelineContext.cpp [hunk 2]
```diff
@@ -377,6 +377,13 @@ void PipelineContext::setOptionsInPipeline(Pipeline *pipeline) const {
       shaderOptions.updateDescInElf = shaderInfo->options.updateDescInElf;
       shaderOptions.unrollThreshold = shaderInfo->options.unrollThreshold;
 
+      static_assert(static_cast<lgc::DenormalMode>(Vkgc::DenormalMode::Auto) == lgc::DenormalMode::Auto, "mismatch");
```
Comment (amdrexu) [inline llpc/context/llpcPipelineContext.cpp:380]: -> Mismatch, Uppercase initial. 
Apply it to following codes.

File: lgc/interface/lgc/Pipeline.h [hunk 3]
```diff
@@ -81,6 +81,13 @@ enum class NggSubgroupSizing : unsigned {
                     ///  primsPerSubgroup
 };
 
+/// Enumerate denormal override modes.
+enum class DenormalMode : unsigned {
```
Comment (amdrexu) [inline lgc/interface/lgc/Pipeline.h:85]: No sure if you need to pass this option to LGC. Currently, only front-end uses this option.

File: llpc/context/llpcPipelineContext.cpp (hunk 2 above)
Comment (perlfu) [reply llpc/context/llpcPipelineContext.cpp:380]: I am happy to change this, but every other static assertion in this file uses lowercase "mismatch".
Can I leave this change with it lowercase and then submit a follow up PR to rewrite them all to uppercase "Mismatch" if that is your preference?

Comment (amdrexu) [reply llpc/context/llpcPipelineContext.cpp:380]: That's OK. Keep it consistent with the environment. Then change all.

File: lgc/interface/lgc/Pipeline.h (hunk 3 above)
Comment (perlfu) [reply lgc/interface/lgc/Pipeline.h:85]: llpcPipelineContext.cpp is based on the lgc types, so I think this is unavoidable?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 23bd60054cd58f78eba18f87388b8da209411ecf
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/Pipeline.h (hunk 3 above)
Comment (amdrexu) [reply lgc/interface/lgc/Pipeline.h:85]: llpcPipelineContext less depends on LGC types now. LGC has its own LgcContext. I am OK if you keep this code.

**Pull request merged**